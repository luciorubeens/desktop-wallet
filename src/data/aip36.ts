export const EntitySourceControl = [
	{
		value: "bitbucket",
		name: "BitBucket",
		url: "https://bitbucket.org",
	},
	{
		value: "github",
		name: "GitHub",
		url: "https://github.com",
	},
	{
		value: "gitlab",
		name: "GitLab",
		url: "https://gitlab.com",
	},
	{
		value: "npm",
		name: "NPM",
		url: "https://npmjs.com",
	},
];

export const EntitySocialMedia = [
	{
		value: "discord",
		name: "Discord",
		url: "https://discord.com",
	},
	{
		value: "facebook",
		name: "Facebook",
		url: "https://facebook.com",
	},
	{
		value: "instagram",
		name: "Instagram",
		url: "https://instagram.com",
	},
	{
		value: "linkedin",
		name: "LinkedIn",
		url: "https://linkedin.com",
	},
	{
		value: "medium",
		name: "Medium",
		url: "https://medium.com",
	},
	{
		value: "reddit",
		name: "Reddit",
		url: "https://reddit.com",
	},
	{
		value: "slack",
		name: "Slack",
		url: "https://slack.com",
	},
	{
		value: "telegram",
		name: "Telegram",
		url: "https://telegram.org",
	},
	{
		value: "twitter",
		name: "Twitter",
		url: "https://twitter.com",
	},
	{
		value: "wechat",
		name: "WeChat",
		url: "https://wechat.com",
	},
	{
		value: "youtube",
		name: "YouTube",
		url: "https://youtube.com",
	},
];

export const EntityMedia = [
	{
		type: "image",
		value: "flickr",
		name: "Flickr",
		url: "https://flickr.com",
	},
	{
		type: "image",
		value: "imgur",
		name: "Imgur",
		url: "https://imgur.com",
	},
	{
		type: "video",
		value: "youtube",
		name: "YouTube",
		url: "https://youtube.com",
	},
	{
		type: "video",
		value: "vimeo",
		name: "Vimeo",
		url: "https://vimeo.com",
	},
];
