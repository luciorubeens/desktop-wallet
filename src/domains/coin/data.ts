interface CoinExtendedManifestStruct {
	borderClass: string;
	textClass: string;
	iconName: string;
	networks?: Record<string, CoinExtendedManifestStruct>;
}

export const CoinExtendedManifest: Record<string, CoinExtendedManifestStruct> = {
	ARK: {
		borderClass: "border-theme-danger-light",
		textClass: "text-theme-danger-400",
		iconName: "Ark",
		networks: {
			Devnet: {
				borderClass: "border-theme-primary-100",
				textClass: "text-theme-primary-400",
				iconName: "Ark",
			},
		},
	},
	Bitcoin: {
		borderClass: "border-theme-warning-200",
		textClass: "text-theme-warning-400",
		iconName: "Bitcoin",
	},
	Ethereum: {
		borderClass: "border-theme-neutral-600",
		textClass: "text-theme-neutral-800",
		iconName: "Ethereum",
	},
	Lisk: {
		borderClass: "border-theme-primary-400",
		textClass: "text-theme-primary",
		iconName: "Lisk",
	},
	Ripple: {
		borderClass: "border-theme-primary-500",
		textClass: "text-theme-primary-dark",
		iconName: "Ripple",
	},
};
