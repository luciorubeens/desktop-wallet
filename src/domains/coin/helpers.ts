import { coinsExtendedManifest } from "./data";

export const getCoinExtendedManifest = ({ coin, network }: { coin: string; network?: string }) => {
	const coinExtendManifest = coinsExtendedManifest[coin];

	if (!coinExtendManifest) {
		return;
	}

	const networkManifest = network ? coinExtendManifest?.networks?.[network] : undefined;
	const manifest = Object.assign({}, coinExtendManifest, networkManifest);

	return manifest;
};
