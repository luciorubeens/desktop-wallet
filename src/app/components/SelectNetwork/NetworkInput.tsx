import { CoinIcon } from "domains/coin/components/CoinIcon";
import { CoinType } from "domains/coin/models";
import React from "react";

import { Input, InputAddon, InputAddonStart, InputGroup } from "../Input";

type Props = {
	network?: CoinType;
	suggestion?: string;
} & React.InputHTMLAttributes<any>;

const TypeSuggestion = ({ value }: { value?: string }) => {
	return (
		<InputAddon className="border border-transparent pl-15 py-3 opacity-50 pointer-events-none">{value}</InputAddon>
	);
};

export const NetworkInput = React.forwardRef<HTMLInputElement, Props>(
	({ network, suggestion, ...props }: Props, ref) => {
		return (
			<InputGroup>
				<InputAddonStart className="px-4">
					<CoinIcon coin={network?.coin || ""} network={network?.network} size="sm" showTooltip={false} />
				</InputAddonStart>
				<TypeSuggestion value={suggestion} />
				<Input ref={ref} className="pl-15" {...props} />
			</InputGroup>
		);
	},
);

NetworkInput.displayName = "NetworkInput";
