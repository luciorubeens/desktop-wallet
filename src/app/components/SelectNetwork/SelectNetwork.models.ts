import { CoinType } from "domains/coin/models";

export type Network = CoinType & { displayName: string };
