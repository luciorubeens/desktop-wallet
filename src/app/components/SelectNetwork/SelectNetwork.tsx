import { useEnvironment } from "app/contexts";
import { CoinIcon } from "domains/coin/components/CoinIcon";
import { CoinType } from "domains/coin/models";
import Downshift from "downshift";
import React from "react";
import { styled } from "twin.macro";

type SelectNetworkProps = {
	networks: CoinType[];
	placeholder?: string;
	name?: string;
	value?: string;
	onSelect?: (network: any) => void;
};

type InputValue = any;

const TypeAhead = ({ input, matches }: any) => {
	const formatTypeHeadToMatchInputCase = (network: CoinType, input: InputValue) => {
		return [input, network?.coin.slice(input.length)].join("");
	};

	const typeaheadFormatted = matches.length === 1 ? formatTypeHeadToMatchInputCase(matches[0], input) : "";

	return (
		<div className="relative z-10 w-full" data-testid={`select-network__typeahead-${typeaheadFormatted}`}>
			<div className="absolute left-0 font-semibold top-2 text-theme-neutral-light">{typeaheadFormatted}</div>
		</div>
	);
};

const InputWrapper = styled.div`
	&:hover {
		box-shadow: 0 0 0 1px var(--theme-color-primary);
	}
`;

export const SelectNetwork = ({ networks, placeholder, onSelect, name }: SelectNetworkProps) => {
	const isMatch = (network: CoinType, input: InputValue) => {
		if (!input) {
			return false;
		}

		return network.coin.toLowerCase().startsWith(input.toLowerCase());
	};

	const getMatches = (networks: CoinType[], input: InputValue) => {
		return networks.filter((network: CoinType) => isMatch(network, input));
	};

	const assetClassName = (network: CoinType, selectedAsset?: CoinType, input?: InputValue) => {
		// Selected is me. Show me green
		if (selectedAsset && selectedAsset.coin === network.coin && selectedAsset.network === network.network) {
			return "text-theme-success-500 border-theme-success-200";
		}

		// Selection is made but not me. Show me disabled
		if (selectedAsset && selectedAsset.coin !== network.coin && selectedAsset.network !== network.network) {
			return "text-theme-neutral-light";
		}

		// Initial state. Nothing entered, nothing selected
		if (!input) {
			return undefined;
		}

		// Input entered, matching with input. Show normal colors
		if (isMatch(network, input)) {
			return undefined;
		}

		// Disabled otherwise
		return "text-theme-neutral-light";
	};

	return (
		<Downshift itemToString={(i) => i?.coin} onSelect={onSelect}>
			{({
				getLabelProps,
				getInputProps,
				getItemProps,
				getMenuProps,
				selectItem,
				inputValue,
				selectedItem,
				clearSelection,
			}) => (
				<div className="relative">
					<label {...getLabelProps()} />
					<div className="relative flex items-center w-full flex-inline">
						<InputWrapper className="flex w-full border rounded transition-colors duration-200 bg-theme-background border-theme-neutral-300 hover:outline-none hover:border-theme-primary">
							<div className="px-4 py-2 flex-0 w-14">
								<CoinIcon
									data-testid="select-asset__selected"
									showTooltip={false}
									size="sm"
									coin={selectedItem?.coin}
									network={selectedItem?.name}
								/>
							</div>
							<div className="relative flex-1 p-1 font-semibold text-theme-neutral-800">
								<TypeAhead input={inputValue} matches={getMatches(networks, inputValue)} />
								<input
									name={name}
									{...getInputProps({
										value: inputValue || "",
										placeholder,
										onKeyDown: (event: any) => {
											if (event.key === "Tab" || event.key === "Enter") {
												// Selected if exact match
												const matches = getMatches(networks, inputValue);
												if (matches.length === 1) {
													selectItem(matches[0]);
												}
												event.preventDefault();
												return;
											}

											// Clear selection when user is changing input,
											// and input does is not exact match with previously selected item
											if (selectedItem && selectedItem.name !== inputValue) {
												clearSelection();
											}
										},
									})}
									data-testid="select-asset__input"
									className="relative z-20 w-full h-full font-semibold bg-transparent outline-none"
								/>
							</div>
						</InputWrapper>
					</div>
					{networks && networks.length > 0 && (
						<div data-testid="select-asset__items" className="select-asset__items" {...getMenuProps()}>
							{networks.map((network: CoinType, index: number) => {
								return (
									<div
										key={index}
										className="inline-block pt-6 mr-6 cursor-pointer"
										{...getItemProps({ item: network })}
									>
										<CoinIcon
											iconSize={26}
											size="xl"
											network={network.network}
											coin={network.coin}
											className={assetClassName(network, selectedItem, inputValue)}
										/>
									</div>
								);
							})}
						</div>
					)}
				</div>
			)}
		</Downshift>
	);
};

export const SelectNetworkContainer = (props: SelectNetworkProps) => {
	const env = useEnvironment();
	const networks = React.useMemo(() => env?.availableNetworks(), [env]);

	return <SelectNetwork {...props} networks={networks} />;
};

SelectNetwork.defaultProps = {
	networks: [],
	placeholder: "Enter a network name",
};
