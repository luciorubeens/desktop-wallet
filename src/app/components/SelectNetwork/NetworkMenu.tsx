import { CoinIcon } from "domains/coin/components/CoinIcon";
import { availableNetworksMock } from "domains/coin/data";
import { getCoinExtendedManifest } from "domains/coin/helpers";
import { CoinType } from "domains/coin/models";
import { useCombobox } from "downshift";
import React from "react";

import { NetworkInput } from "./NetworkInput";

const networks = availableNetworksMock;

type CoinWithDisplayName = CoinType & { displayName?: string };

export const NetworkMenu = () => {
	const [value, setValue] = React.useState<string | undefined>(undefined);

	const [items] = React.useState(() =>
		networks.map((network) => {
			const manifest = getCoinExtendedManifest({ coin: network.coin, network: network.network });
			return { ...network, displayName: manifest?.displayName };
		}),
	);

	const {
		selectItem,
		selectedItem,
		getInputProps,
		getMenuProps,
		getComboboxProps,
		getItemProps,
		highlightedIndex,
		setHighlightedIndex,
	} = useCombobox<CoinWithDisplayName | null>({
		items,
		itemToString: (item) => item?.displayName || "",
		onInputValueChange: ({ inputValue }) => {
			if (!inputValue) {
				selectItem(null);
			}
			setValue(inputValue);
		},
	});

	const inputMatchesIndex = React.useMemo(() => {
		if (!value) {
			return [];
		}

		return items
			.map((network, index) => network.displayName?.toLowerCase().startsWith(value.toLowerCase().trim()) && index)
			.filter((i): i is number => {
				return typeof i === "number";
			});
	}, [value, items]);

	const isExactMatch = React.useMemo(() => inputMatchesIndex.length === 1, [inputMatchesIndex]);
	const highlightedItem = React.useMemo(() => (inputMatchesIndex.length ? items[inputMatchesIndex[0]] : undefined), [
		items,
		inputMatchesIndex,
	]);

	React.useEffect(() => {
		if (inputMatchesIndex.length > 0) {
			setHighlightedIndex(inputMatchesIndex[0]);
		}
	}, [inputMatchesIndex, setHighlightedIndex]);

	return (
		<>
			<div {...getComboboxProps()}>
				<NetworkInput
					network={selectedItem}
					suggestion={isExactMatch ? highlightedItem?.displayName : undefined}
					{...getInputProps()}
				/>
			</div>
			<ul {...getMenuProps()}>
				{items.map((item, index) => (
					<li
						className="inline-block pt-6 mr-6 cursor-pointer"
						key={index}
						{...getItemProps({
							item,
							index,
							disabled:
								!selectedItem && inputMatchesIndex.length > 0 && !inputMatchesIndex.includes(index),
						})}
					>
						<CoinIcon network={item.network} coin={item.coin} size="xl" iconSize={26} />
					</li>
				))}
			</ul>
		</>
	);
};
