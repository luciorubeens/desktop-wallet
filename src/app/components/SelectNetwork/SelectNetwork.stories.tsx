import React from "react";

import { SelectNetwork } from "./SelectNetwork";

export default {
	title: "App / Components / SelectNetwork",
};

export const Default = () => {
	const networks = [
		{
			coin: "ARK",
			name: "ARK Ecosystem",
		},
		{
			coin: "Bitcoin",
			name: "Bitcoin",
		},
		{
			coin: "Ethereum",
			name: "Ethereum",
		},
		{
			coin: "Lisk",
			name: "Lisk",
		},
		{
			coin: "Ripple",
			name: "Ripple",
		},
	];

	return (
		<div className="p-5 w-128">
			<SelectNetwork networks={networks} />
		</div>
	);
};
