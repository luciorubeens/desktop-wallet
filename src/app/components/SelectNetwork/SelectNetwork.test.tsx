import { availableNetworksMock } from "domains/coin/data";
import React from "react";
import { act, fireEvent, render } from "utils/testing-library";

import { SelectNetwork } from "./SelectNetwork";

describe("SelectNetwork", () => {
	it("should render", () => {
		const { container } = render(<SelectNetwork />);
		expect(container).toMatchSnapshot();
	});

	it("should render with networks", () => {
		const { container } = render(<SelectNetwork networks={availableNetworksMock} />);
		expect(container).toMatchSnapshot();
	});

	it("should show typeahead when typing has found one exact match", () => {
		const { getByTestId } = render(<SelectNetwork networks={availableNetworksMock} />);
		const input = getByTestId("select-asset__input");
		act(() => {
			fireEvent.change(input, { target: { value: "Bitco" } });
		});

		expect(getByTestId("select-network__typeahead-Bitcoin")).toBeTruthy();
	});

	it("should select first matching asset with enter", () => {
		const { getByTestId } = render(<SelectNetwork networks={availableNetworksMock} />);
		const input = getByTestId("select-asset__input");
		act(() => {
			fireEvent.change(input, { target: { value: "Bitco" } });
		});

		act(() => {
			fireEvent.keyDown(input, { key: "Enter", code: 13 });
		});

		expect(getByTestId("select-asset__selected")).toHaveAttribute("aria-label", "Bitcoin");
	});

	it("should select first matching asset with tab", () => {
		const { getByTestId } = render(<SelectNetwork networks={availableNetworksMock} />);
		const input = getByTestId("select-asset__input");
		act(() => {
			fireEvent.change(input, { target: { value: "Bitcoi" } });
		});

		act(() => {
			fireEvent.keyDown(input, { key: "Tab", code: 9 });
		});

		expect(getByTestId("select-asset__selected")).toHaveAttribute("aria-label", "Bitcoin");
	});

	it("should not select non-matching asset after key input and tab", () => {
		const { getByTestId, queryByTestId } = render(<SelectNetwork networks={availableNetworksMock} />);
		const input = getByTestId("select-asset__input");
		act(() => {
			fireEvent.change(input, { target: { value: "Bot" } });
		});

		act(() => {
			fireEvent.keyDown(input, { key: "Tab", code: 9 });
		});

		expect(queryByTestId("select-asset__selected-Bitcoin")).not.toBeTruthy();
	});

	it("should not select first matched asset after random key enter", () => {
		const { getByTestId, queryByTestId } = render(<SelectNetwork networks={availableNetworksMock} />);
		const input = getByTestId("select-asset__input");

		act(() => {
			fireEvent.change(input, { target: { value: "Bitco" } });
		});

		act(() => {
			fireEvent.keyDown(input, { key: "A", code: 65 });
		});

		expect(queryByTestId("select-asset__selected-Bitcoin")).not.toBeTruthy();
	});

	it("should clear selection when changing input", () => {
		const { getByTestId, queryByTestId } = render(<SelectNetwork networks={availableNetworksMock} />);
		const input = getByTestId("select-asset__input");

		act(() => {
			fireEvent.change(input, { target: { value: "Bitcoin" } });
		});

		act(() => {
			fireEvent.keyDown(input, { key: "Enter", code: 13 });
		});

		expect(getByTestId("select-asset__selected")).toBeTruthy();

		act(() => {
			fireEvent.change(input, { target: { value: "test" } });
		});

		act(() => {
			fireEvent.keyDown(input, { key: "A", code: 65 });
		});
		act(() => {
			fireEvent.keyDown(input, { key: "B", code: 65 });
		});

		expect(queryByTestId("select-asset__selected")).not.toHaveAttribute("aria-label");
	});
});
