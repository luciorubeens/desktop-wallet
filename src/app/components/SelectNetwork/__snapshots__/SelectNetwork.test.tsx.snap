// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectNetwork should render 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    class="relative"
    role="combobox"
  >
    <label
      for="downshift-0-input"
      id="downshift-0-label"
    />
    <div
      class="relative flex items-center w-full flex-inline"
    >
      <div
        class="sc-AxiKw fRbeDg flex w-full border rounded transition-colors duration-200 bg-theme-background border-theme-neutral-300 hover:outline-none hover:border-theme-primary"
      >
        <div
          class="px-4 py-2 flex-0 w-14"
        >
          <div
            class="sc-AxjAm IjnG border-theme-neutral-light text-theme-neutral"
            data-testid="select-asset__selected"
          />
        </div>
        <div
          class="relative flex-1 p-1 font-semibold text-theme-neutral-800"
        >
          <div
            class="relative z-10 w-full"
            data-testid="select-network__typeahead-"
          >
            <div
              class="absolute left-0 font-semibold top-2 text-theme-neutral-light"
            />
          </div>
          <input
            aria-autocomplete="list"
            aria-labelledby="downshift-0-label"
            autocomplete="off"
            class="relative z-20 w-full h-full font-semibold bg-transparent outline-none"
            data-testid="select-asset__input"
            id="downshift-0-input"
            placeholder="Enter a network name"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SelectNetwork should render with networks 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-1-label"
    class="relative"
    role="combobox"
  >
    <label
      for="downshift-1-input"
      id="downshift-1-label"
    />
    <div
      class="relative flex items-center w-full flex-inline"
    >
      <div
        class="sc-AxiKw fRbeDg flex w-full border rounded transition-colors duration-200 bg-theme-background border-theme-neutral-300 hover:outline-none hover:border-theme-primary"
      >
        <div
          class="px-4 py-2 flex-0 w-14"
        >
          <div
            class="sc-AxjAm IjnG border-theme-neutral-light text-theme-neutral"
            data-testid="select-asset__selected"
          />
        </div>
        <div
          class="relative flex-1 p-1 font-semibold text-theme-neutral-800"
        >
          <div
            class="relative z-10 w-full"
            data-testid="select-network__typeahead-"
          >
            <div
              class="absolute left-0 font-semibold top-2 text-theme-neutral-light"
            />
          </div>
          <input
            aria-autocomplete="list"
            aria-labelledby="downshift-1-label"
            autocomplete="off"
            class="relative z-20 w-full h-full font-semibold bg-transparent outline-none"
            data-testid="select-asset__input"
            id="downshift-1-input"
            placeholder="Enter a network name"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      aria-labelledby="downshift-1-label"
      class="select-asset__items"
      data-testid="select-asset__items"
      id="downshift-1-menu"
      role="listbox"
    >
      <div
        aria-selected="false"
        class="inline-block pt-6 mr-6 cursor-pointer"
        id="downshift-1-item-0"
        role="option"
      >
        <div
          aria-label="Ark"
          class="sc-AxjAm eZoLDa border-theme-neutral-light text-theme-neutral"
          data-testid="CoinIcon"
        />
      </div>
      <div
        aria-selected="false"
        class="inline-block pt-6 mr-6 cursor-pointer"
        id="downshift-1-item-1"
        role="option"
      >
        <div
          aria-label="Bitcoin"
          class="sc-AxjAm eZoLDa border-theme-warning-200 text-theme-warning-400"
          data-testid="CoinIcon"
        >
          <div
            class="sc-AxirZ hyqZof"
            data-testid="CoinIcon__icon"
            height="26"
            width="26"
          >
            <svg>
              btc.svg
            </svg>
          </div>
        </div>
      </div>
      <div
        aria-selected="false"
        class="inline-block pt-6 mr-6 cursor-pointer"
        id="downshift-1-item-2"
        role="option"
      >
        <div
          aria-label="Ethereum"
          class="sc-AxjAm eZoLDa border-theme-neutral-600 text-theme-neutral-800"
          data-testid="CoinIcon"
        >
          <div
            class="sc-AxirZ hyqZof"
            data-testid="CoinIcon__icon"
            height="26"
            width="26"
          >
            <svg>
              eth.svg
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
